README.md
# ProjectSpaceInvaders


CHANGELOG.md
# Changelog for `ProjectSpaceInvaders`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/Main.hs
module Main (main) where

import UnitTests
import PropertiesTests
import Test.QuickCheck
import Test.HUnit

main :: IO ()
main = do
  runTestTT unitTestReplayGame
  runTestTT unitTestAddLetter
  runTestTT unitTestDeleteLastLetter
  runTestTT unitTestStopPlayerMoving
  runTestTT unitTestChangePlayerVelocityRight
  runTestTT unitTestChangePlayerVelocityLeft
  runTestTT unitTestSortScores
  runTestTT unitTestChangePlayerScoreSaved
  runTestTT unitTestChangePlayerHighscore
  runTestTT unitTestEnemyTooFar
  runTestTT unitTestCheckEndGame
  runTestTT unitTestChangePlayerAlive
  runTestTT unitTestBulletKillPlayer
  runTestTT unitTestEnemiesKillPlayer
  runTestTT unitTestChangeEnemyAlive
  runTestTT unitTestBulletKillEnemy
  runTestTT unitTestBulletsKillEnemy
  runTestTT unitTestBulletsKillEnemies
  runTestTT unitTestUsedBullet
  runTestTT unitTestUsedPlayerBullets
  runTestTT unitTestAddPoints
  runTestTT unitTestPlayerKillEnemies
  runTestTT unitTestPlayerOutOfScope
  runTestTT unitTestUpdatePlayerPosition
  runTestTT unitTestUpdateBulletPosition
  runTestTT unitTestBulletOutOfScope
  runTestTT unitTestAddEnemiesBullets
  runTestTT unitTestMoveEnemyRight
  runTestTT unitTestMoveEnemyLeft
  runTestTT unitTestMoveEnemyDown
  runTestTT unitTestUpdateEnemyPosition
  runTestTT unitTestUpdateEnemiesPosition
  runTestTT unitTestGetNextDirection
  runTestTT unitTestShouldChangeDirection
  runTestTT unitTestNextDirection
  runTestTT unitTestSaveScore
  runTestTT unitTestReadFromFile
  runTestTT unitTestLoadHighScore
  runTestTT unitTestLoadPlayerHighScore
  runTestTT unitTestLoadHighScores
  quickCheck testPlayerShoot
  quickCheck testBulletsKillEnemies
  quickCheck testRemoveUsedBullets
  quickCheck testPlayerKillEnemies
  quickCheck testUpdatePlayerPosition
  quickCheck testBulletOutOfScope
  quickCheck testRemoveBulletsOutOfScope
  quickCheck testUpdateBulletsPosition
  quickCheck testSelectRandomEnemies
  quickCheck testEnemiesShoot
  quickCheck testGenerateNewEnemies


test/PropertiesTests.hs
module PropertiesTests where 

import UpdateGame
import System.Random 
import Test.QuickCheck

instance Arbitrary Bullet where
  arbitrary = do
    posX <- fromIntegral <$> chooseInt (-500, 500)
    posY <- fromIntegral <$> chooseInt (-300, 300)
    let velX = 0
    velY <- elements [-2, -2.5, -3, -3.5, -4, -4.5, -5, -5.5, -6, -6.5, 0, 3]
    used <- arbitrary
    return (Bullet (posX, posY) (velX, velY) used)

instance Arbitrary Player where
  arbitrary = do
    posX <- fromIntegral <$> chooseInt (-475, 475)
    let posY = -250
    velX <- elements [-4, 0, 4]
    let velY = 0
    let alive = True 
    name <- arbitrary
    points <- choose (0, 1000)
    highscore <- choose (0, 2000)
    scoreSaved <- arbitrary
    return (Player (posX, posY) (velX, velY) alive name points highscore scoreSaved)

instance Arbitrary Enemy where 
  arbitrary = do
    posX <- fromIntegral <$> chooseInt (-480, 480)
    posY <- fromIntegral <$> chooseInt (-250,280)
    let alive = True 
    return (Enemy (posX,posY) alive)

instance Arbitrary EnemyDirection where
  arbitrary = elements [LeftDir, RightDir, DownDirAfterRight, DownDirAfterLeft]

instance Arbitrary GameState where
  arbitrary = do
    let menu = GameMenu
    player <- arbitrary
    playerBullets <- arbitrary
    generalHighscore <- arbitrary
    bestScores <- arbitrary
    let fileScores = "scores.txt"
    enemies <- arbitrary
    newEnemiesTimer <- chooseInt (0, 200)
    enemyDirection <- arbitrary
    let generator = mkStdGen 0
    enemiesBullets <- arbitrary
    shootTimer <- chooseInt (0, 41)
    countOfRound <- fromIntegral <$> chooseInt (0, 200)
    let exit = False
    return (Game menu player playerBullets generalHighscore bestScores fileScores enemies newEnemiesTimer enemyDirection generator enemiesBullets shootTimer countOfRound exit)

testPlayerShoot :: GameState -> Bool 
testPlayerShoot game = length (playerBullets updatedGame) == length initialBullets + 1
    && bulletPos newBullet == playerPos (player game)
    && bulletVel newBullet == (0, 3)
    && not (bulletUsed newBullet)
    && all (`elem` playerBullets updatedGame) initialBullets 
    where 
        initialBullets = playerBullets game
        updatedGame = playerShoot game
        newBullet = head (playerBullets updatedGame)

testBulletsKillEnemies :: [Bullet] -> [Enemy] -> Bool
testBulletsKillEnemies bullets enemies =  length (bulletsKillEnemies bullets enemies) == length enemies 
    && all (\enemy -> not (bulletsKillEnemy bullets enemy) ) (filter enemyAlive (bulletsKillEnemies bullets enemies))
    && all (\enemy -> bulletsKillEnemy bullets enemy) (filter (\enemy -> not (enemyAlive enemy)) (bulletsKillEnemies bullets enemies))

testRemoveUsedBullets :: [Bullet] -> Bool
testRemoveUsedBullets bullets = all (\bullet -> not (bulletUsed bullet)) filteredBullets && all (\bullet -> bullet `elem` filteredBullets) (filter (not . bulletUsed) bullets)
    where
        filteredBullets = removeUsedBullets bullets 

testPlayerKillEnemies :: GameState -> Bool
testPlayerKillEnemies game = enemiesAfter + killedEnemies == enemiesBefore && bulletsAfter + usedBullets ==  bulletsBefore && pointsAfter == pointsBefore + killedEnemies
  where 
    startGame = game {playerBullets = removeUsedBullets (playerBullets game)}
    newGame = playerKillEnemies startGame
    enemiesBefore = length (enemies startGame)
    enemiesAfter = length (enemies newGame)
    killedEnemies = length (filter (not . enemyAlive) (bulletsKillEnemies (playerBullets startGame) (enemies startGame)))
    bulletsBefore = length (playerBullets startGame)
    bulletsAfter = length (playerBullets newGame)
    usedBullets = length (filter bulletUsed (usedPlayerBullets (enemies startGame) (playerBullets startGame)))
    pointsBefore = playerPoints (player startGame)
    pointsAfter = playerPoints (player newGame)

testUpdatePlayerPosition :: Player -> Bool
testUpdatePlayerPosition player = posX >= -478 && posX <= 478 && posY == -250 
    where 
        updatedPlayer = updatePlayerPosition player
        (posX, posY) = playerPos updatedPlayer

testBulletOutOfScope :: Bullet -> Bool
testBulletOutOfScope bullet = (snd (bulletPos newBullet) > 300 || snd (bulletPos newBullet) < -300 || fst (bulletPos newBullet) > 500 || fst (bulletPos newBullet) < -500) == bulletOutOfScope newBullet
    where 
        (posX, posY) = bulletPos newBullet
        newBullet = updateBulletPosition bullet 

testRemoveBulletsOutOfScope :: [Bullet] -> Bool
testRemoveBulletsOutOfScope bullets = all (\bullet -> not (bulletOutOfScope bullet)) filteredBullets && all (\bullet -> bullet `elem` filteredBullets) newBullets
    where 
        filteredBullets = removeBulletsOutOfScope bullets
        outOfScopeBullets = filter bulletOutOfScope bullets
        newBullets = filter (\bullet -> bullet `notElem` outOfScopeBullets) bullets

testUpdateBulletsPosition :: [Bullet] -> Bool
testUpdateBulletsPosition bullets = updatedBullets == expectedBullets 
    where  
        updatedBullets = updateBulletsPosition bullets
        expectedBullets = filter (\bullet -> (snd (bulletPos bullet) <= 300 && snd (bulletPos bullet) >= -300 && fst (bulletPos bullet) <= 500 && fst (bulletPos bullet) >= -500)) (map updateBulletPosition bullets)

testSelectRandomEnemies :: [Enemy] -> Bool
testSelectRandomEnemies enemies = selectedCount == count && all (`elem` enemies) selected 
    where
        count = min 4 (length enemies) 
        (selected, _) = selectRandomEnemies (mkStdGen 0) enemies count
        selectedCount = length selected

testEnemiesShoot :: GameState -> Bool
testEnemiesShoot game = all (\bullet -> bullet `elem` newBullets) oldBullets 
  && (if oldTimer == 41 then newTimer == 0 else newTimer == oldTimer + 1) 
  && (if oldTimer == 0 then length newBullets == (length oldBullets) + count else length newBullets == length oldBullets)
  where 
    result = enemiesShoot game
    newBullets = enemiesBullets result
    oldBullets = enemiesBullets game 
    newTimer = shootTimer result
    oldTimer = shootTimer game 
    count = min 4 (length (enemies game))

testGenerateNewEnemies :: GameState -> Bool 
testGenerateNewEnemies game 
  | enemies game /= [] = enemies game == enemies newGame
  | newEnemiesTimer game < 200 = newTimer == oldTimer + 1 && enemies game == enemies newGame 
  | otherwise = enemies newGame == seriesOfEnemies && newRound == oldRound + 1 && newTimer == 0 
    where
      newGame = generateNewEnemies game
      newTimer = newEnemiesTimer newGame
      oldTimer = newEnemiesTimer game 
      newRound = countOfRound newGame
      oldRound = countOfRound game 

src/Game.hs
module Game where 

import HandleGame
import DrawGame 
import UpdateGame
import Graphics.Gloss.Interface.IO.Game 
import Graphics.Gloss.Juicy
import System.Random 
import System.Exit (exitWith, ExitCode(..))

window :: Display
window = InWindow "Space Invaders" (1000, 600) (10,10)  

playGame :: IO ()
playGame = do
  gen <- getStdGen
  bgPicture <- loadJuicyPNG "bgPicture.png"
  picturePlayer <- loadJuicyPNG "player.png"
  pictureEnemy <- loadJuicyPNG "enemy.png"
  picturePlayerBullet <- loadJuicyPNG "playerBullet.png"
  pictureEnemyBullet <- loadJuicyPNG "enemyBullet.png"
  case (,,,,) <$> bgPicture <*> picturePlayer <*> pictureEnemy <*> picturePlayerBullet <*> pictureEnemyBullet of
    Just (bgPic, playerPic, enemyPic, playerBulletPic, enemyBulletPic) -> do
      playIO window black 60 (initialState gen) (drawGameIO bgPic playerPic enemyPic playerBulletPic enemyBulletPic) handleEventIO update
    _ -> do
      putStrLn "Error loading images"
      exitWith (ExitFailure 1)

src/DrawGame.hs
module DrawGame where 

import UpdateGame
import Graphics.Gloss

drawStartMenu :: Picture -> Picture
drawStartMenu bgPicture = pictures [bgPicture, titleLabel, startLabel, instructionLabel, boardLabel, exitLabel]
  where
    titleLabel = translate (-340) 50 $ color white $ scale 0.75 0.75 $ text "Space Invaders"
    startLabel = translate (-210) (-20) $ color white $ scale 0.2 0.2 $ text "press Enter to start the game"
    instructionLabel = translate (-210) (-70) $ color white $ scale 0.2 0.2 $ text "press i to show the instruction"
    boardLabel = translate (-210) (-120) $ color white $ scale 0.2 0.2 $ text "press x to show the leaderboard"
    exitLabel = translate (-210) (-170) $ color white $ scale 0.2 0.2 $ text "press Esc to exit the game"

drawInstructionMenu :: Picture -> Picture
drawInstructionMenu bgPicture = pictures [bgPicture, titleLabel, ruleFirstLabel, ruleSecondLabel, rightLabel, leftLabel, spaceLabel, startLabel, boardLabel, exitLabel]
  where
    titleLabel = translate (-340) 100 $ color white $ scale 0.75 0.75 $ text "Space Invaders"
    ruleFirstLabel = translate (-230) 50 $ color white $ scale 0.2 0.2 $ text "Kill as many enemies as you can"
    ruleSecondLabel = translate (-280) 0 $ color white $ scale 0.2 0.2 $ text "Don't let them kill you as long as you can"
    rightLabel = translate (-170) (-50) $ color white $ scale 0.2 0.2 $ text "press -> to move right"
    leftLabel = translate (-160) (-100) $ color white $ scale 0.2 0.2 $ text "press <- to move left"
    spaceLabel = translate (-150) (-150) $ color white $ scale 0.2 0.2 $ text "press Space to shoot"
    startLabel = translate (-490) (-250) $ color black $ scale 0.1 0.1 $ text "press Enter to start the game"
    boardLabel = translate (-490) (-270) $ color black $ scale 0.1 0.1 $ text "press x to show the leaderboard"
    exitLabel = translate (-490) (-290) $ color black $ scale 0.1 0.1 $ text "press Esc to exit the game"

drawNickMenu :: Picture -> GameState -> Picture
drawNickMenu bgPicture game = pictures [bgPicture, nameLabel, playLabel]
  where
    nameLabel = translate (-120) 0 $ color white $ scale 0.15 0.15 $ text $ "Enter your name: " ++ playerName (player game)
    playLabel =  translate (-100) (-100) $ color white $ scale 0.15 0.15 $ text "press Enter to play" 

drawStartBoardMenu :: Picture -> GameState -> Picture 
drawStartBoardMenu bgPicture game = pictures $ [bgPicture, startLabel, instructionLabel, exitLabel] ++ (zipWith drawTopTen [0..] (bestScores game))
  where 
    startLabel = translate (-490) (-250) $ color black $ scale 0.1 0.1 $ text "press Enter to start the game"
    instructionLabel = translate (-490) (-270) $ color black $ scale 0.1 0.1 $ text "press i to show the instruction"
    exitLabel = translate (-490) (-290) $ color black $ scale 0.1 0.1 $ text "press Esc to exit the game"

drawBoardMenu :: Picture -> GameState -> Picture 
drawBoardMenu bgPicture game = pictures $ [bgPicture, startLabel, exitLabel] ++ (zipWith drawTopTen [0..] (bestScores game))
  where 
    startLabel = translate (-490) (-270) $ color black $ scale 0.1 0.1 $ text "press Enter to start the game again"
    exitLabel = translate (-490) (-290) $ color black $ scale 0.1 0.1 $ text "press Esc to exit the game"

drawTopTen :: Int -> (String, Int) -> Picture
drawTopTen index (name, points) = translate (-120) (220 - fromIntegral index * 40) $ scale 0.2 0.2 $ color white $ text (show (index + 1) ++ ". " ++ name ++ ": " ++ show points)

drawPlayer :: Picture -> Player -> Picture
drawPlayer picturePlayer player = translate (fst (playerPos player)) (snd (playerPos player)) $ scale 0.4 0.4 picturePlayer 

drawEnemiesBullets :: Picture -> [Bullet] -> Picture
drawEnemiesBullets pictureEnemyBullet bullets = pictures [drawEnemyBullet pictureEnemyBullet bullet | bullet <- bullets]

drawEnemyBullet :: Picture -> Bullet -> Picture
drawEnemyBullet pictureEnemyBullet bullet = translate (fst (bulletPos bullet)) (snd (bulletPos bullet)) $ scale 0.5 0.5 pictureEnemyBullet

drawPlayerBullets :: Picture -> [Bullet] -> Picture
drawPlayerBullets picturePlayerBullet bullets = pictures [drawPlayerBullet picturePlayerBullet bullet | bullet <- bullets]

drawPlayerBullet :: Picture -> Bullet -> Picture
drawPlayerBullet picturePlayerBullet bullet = translate (fst (bulletPos bullet)) (snd (bulletPos bullet)) $ scale 0.2 0.2 picturePlayerBullet

drawEnemies :: Picture -> [Enemy] -> Picture
drawEnemies pictureEnemy enemies = pictures [drawEnemy pictureEnemy enemy | enemy <- enemies]

drawEnemy :: Picture -> Enemy -> Picture
drawEnemy pictureEnemy enemy = translate (fst (enemyPos enemy)) (snd (enemyPos enemy)) $ scale 0.25 0.25 pictureEnemy 

drawPoints :: Int -> Picture
drawPoints points = translate (-490) 260 $ color white $ scale 0.1 0.1 $ text ("Your points: " ++ show points)

drawGameMenu :: Picture -> Picture -> Picture -> Picture -> Picture -> GameState -> Picture 
drawGameMenu bgPicture picturePlayer pictureEnemy picturePlayerBullet pictureEnemyBullet game = pictures [bgPicture, drawPlayerBullets picturePlayerBullet (playerBullets game), drawEnemiesBullets pictureEnemyBullet (enemiesBullets game), drawPlayer picturePlayer (player game), drawEnemies pictureEnemy (enemies game), drawPoints (playerPoints (player game))]

drawEndMenu :: Picture -> GameState -> Picture
drawEndMenu bgPicture game = pictures [ bgPicture, overLabel, playerScoreLabel, playerHighscoreLabel, globalHighscoreLabel, startLabel, boardLabel, exitLabel]
  where 
    overLabel = translate (-210) 50 $ color white $ scale 0.5 0.5 $ text "GAME OVER"
    playerScoreLabel = translate (-205) 0 $ color white $ scale 0.3 0.3 $ text $ "Score: " ++ show (playerPoints (player game))
    playerHighscoreLabel = translate (-205) (-50) $ color white $ scale 0.3 0.3 $ text $ "Your highscore: " ++ show (playerHighscore (player game))
    globalHighscoreLabel = translate (-205) (-100) $ color white $ scale 0.3 0.3 $ text $ "Global highscore: " ++ show (generalHighscore game)
    startLabel = translate (-490) (-250) $ color black $ scale 0.1 0.1 $ text "press Enter to start the game again"
    boardLabel = translate (-490) (-270) $ color black $ scale 0.1 0.1 $ text "press x to show the leaderboard"
    exitLabel = translate (-490) (-290) $ color black $ scale 0.1 0.1 $ text "press Esc to exit the game"

drawGameIO :: Picture -> Picture -> Picture -> Picture -> Picture -> GameState -> IO Picture
drawGameIO bgPicture picturePlayer pictureEnemy picturePlayerBullet pictureEnemyBullet game
  | menu game == StartMenu = return $ drawStartMenu bgPicture 
  | menu game == InstructionMenu = return $ drawInstructionMenu bgPicture
  | menu game == NickMenu = return $ drawNickMenu bgPicture game
  | menu game == BoardMenu = return $ drawBoardMenu bgPicture game 
  | menu game == StartBoardMenu = return $ drawStartBoardMenu bgPicture game 
  | menu game == GameMenu = return $ drawGameMenu bgPicture picturePlayer pictureEnemy picturePlayerBullet pictureEnemyBullet game
  | menu game == EndMenu = return $ drawEndMenu bgPicture game


src/UpdateGame.hs
module UpdateGame where 

import Graphics.Gloss 
import System.Random 
import Data.List (delete, sortBy)
import System.Exit ( exitSuccess ) 
import FileOperations 

data Menu = StartMenu | NickMenu | InstructionMenu | StartBoardMenu | BoardMenu | GameMenu | EndMenu 
  deriving (Eq,Show)

data Player = Player 
  { playerPos :: Point,
    playerVel :: Vector, 
    playerAlive :: Bool,
    playerName :: String, 
    playerPoints :: Int, 
    playerHighscore :: Int, 
    playerScoreSaved :: Bool 
  } deriving (Eq, Show)

data Bullet = Bullet
  { bulletPos :: Point,
    bulletVel :: Vector,
    bulletUsed :: Bool 
  } deriving (Eq, Show)

data Enemy = Enemy 
  { enemyPos :: Point,
    enemyAlive :: Bool
  }
  deriving (Eq, Show)

data EnemyDirection = LeftDir | RightDir | DownDirAfterRight | DownDirAfterLeft 
  deriving (Eq, Show)

data GameState = Game 
  { 
    menu :: Menu,
    player :: Player,
    playerBullets :: [Bullet],
    generalHighscore :: Int,
    bestScores :: [(String, Int)],
    fileScores :: String,
    enemies :: [Enemy],
    newEnemiesTimer :: Int,
    enemyDirection :: EnemyDirection,
    generator :: StdGen,
    enemiesBullets :: [Bullet],
    shootTimer :: Int,
    countOfRound :: Float,
    exit :: Bool
  } deriving (Eq,Show)

seriesOfEnemies :: [Enemy]
seriesOfEnemies = concatMap createRow [2 .. 5]
  where
    createRow y = take 10 [Enemy (fromIntegral x, fromIntegral y * 50) True | x <- [-500 .. 500], x `mod` 60 == 0]

initialState :: StdGen -> GameState
initialState gen = Game
  { 
    menu = StartMenu, 
    player = Player (0, -250) (0,0) True "" 0 0 False,
    playerBullets = [],
    generalHighscore = 0,
    bestScores = [],
    fileScores = "scores.txt",
    enemies = seriesOfEnemies,
    newEnemiesTimer = 0,
    enemyDirection = RightDir,
    generator = gen,
    enemiesBullets = [],
    shootTimer = 0,
    countOfRound = 1,
    exit = False
  }

playerShoot :: GameState -> GameState
playerShoot game = game {playerBullets = newBullet : playerBullets game}
    where 
        newBullet = Bullet {bulletPos = playerPos (player game), bulletVel = (0,3), bulletUsed = False}

replayGame :: GameState -> GameState
replayGame game =  game {
    menu = GameMenu, 
    player = Player (0, -250) (0,0) True (playerName (player game)) 0 (playerHighscore (player game)) False,  
    playerBullets = [],
    enemies = seriesOfEnemies,
    newEnemiesTimer = 0,
    enemyDirection = RightDir,
    enemiesBullets = [],
    shootTimer = 0,
    countOfRound = 1,
    exit = False
    }

addLetter :: String -> Player -> Player 
addLetter string player = player {playerName = (playerName player) ++ string}

deleteLastLetter :: Player -> Player 
deleteLastLetter player | playerName player /= "" = player {playerName = init (playerName player)} | otherwise = player 

stopPlayerMoving :: Player -> Player
stopPlayerMoving player = player {playerVel = (0, 0)}

changePlayerVelocityRight :: Player -> Player
changePlayerVelocityRight player 
    | fst (playerPos player) >= 475 = stopPlayerMoving player 
    | otherwise = player {playerVel = (4, 0)}

changePlayerVelocityLeft :: Player -> Player
changePlayerVelocityLeft player 
    | fst (playerPos player) <= -475 = stopPlayerMoving player 
    | otherwise = player {playerVel = (-4, 0)}

sortScores :: [(String, Int)] -> [(String, Int)]
sortScores = sortBy compareSecond
  where compareSecond (_, x) (_, y)
          | x < y = GT
          | x > y = LT
          | otherwise = EQ

loadTopTen :: GameState -> IO GameState
loadTopTen game = do 
    topScores <- loadHighScores (fileScores game) 
    let topTen = take (min 10 (length topScores)) $ sortScores topScores 
    return game {bestScores = topTen}

changePlayerScoreSaved :: Player -> Player 
changePlayerScoreSaved player = player {playerScoreSaved = True}

changePlayerHighscore :: Int -> Player -> Player
changePlayerHighscore points player = player {playerHighscore = points}

saveAndLoadScores :: GameState -> IO GameState
saveAndLoadScores game = do
    saveScore (fileScores game) (playerName (player game)) (playerPoints (player game)) 
    maxScore <- loadHighScore (fileScores game) 
    maxUserScore <- loadPlayerHighScore (playerName (player game)) (fileScores game) 
    topScores <- loadHighScores (fileScores game) 
    let topTen = take (min 10 (length topScores)) $ sortScores topScores 
    let newPlayer = changePlayerScoreSaved (player game)
    return game {player = changePlayerHighscore maxUserScore newPlayer, generalHighscore = maxScore, bestScores = topTen}

enemyTooFar :: Enemy -> Bool 
enemyTooFar enemy = snd (enemyPos enemy) <= -240 

checkEndGame :: GameState -> GameState
checkEndGame game 
    | (playerAlive (player game)) == False || any enemyTooFar (enemies game) = game {menu = EndMenu} 
    | otherwise = game 

changePlayerAlive :: Player -> Player
changePlayerAlive player = player {playerAlive = False}

bulletKillPlayer :: Player -> Bullet -> Bool 
bulletKillPlayer player bullet = fst (bulletPos bullet) <= (posX + 35) && fst (bulletPos bullet) >= posX - 35 && snd (bulletPos bullet) <= (posY + 10) && snd (bulletPos bullet) >= posY - 10
    where
        posX = fst $ playerPos player 
        posY = snd $ playerPos player 

enemiesKillPlayer :: GameState -> GameState
enemiesKillPlayer game 
    | any (bulletKillPlayer (player game)) (enemiesBullets game) = game {player = changePlayerAlive (player game)} 
    | otherwise = game 

changeEnemyAlive :: Enemy -> Enemy
changeEnemyAlive enemy = enemy {enemyAlive = False}

bulletKillEnemy :: Enemy -> Bullet -> Bool
bulletKillEnemy enemy bullet = fst (bulletPos bullet) <= (posX + 20) && fst (bulletPos bullet) >= (posX - 20) && snd (bulletPos bullet) <= (posY + 15) && snd (bulletPos bullet) >= (posY - 15)
    where 
        posX = fst $ enemyPos enemy
        posY = snd $ enemyPos enemy

bulletsKillEnemy :: [Bullet] -> Enemy -> Bool
bulletsKillEnemy bullets enemy = any (bulletKillEnemy enemy) bullets

bulletsKillEnemies :: [Bullet] -> [Enemy] -> [Enemy]
bulletsKillEnemies bullets enemies = fmap (\x -> if bulletsKillEnemy bullets x then changeEnemyAlive x else x) enemies

usedBullet :: [Enemy] -> Bullet -> Bullet
usedBullet enemies bullet 
    | any (\x -> bulletKillEnemy x bullet) enemies = bullet {bulletUsed = True} 
    | otherwise = bullet 

usedPlayerBullets :: [Enemy] -> [Bullet] -> [Bullet]
usedPlayerBullets enemies bullets = fmap (usedBullet enemies) bullets

removeUsedBullets :: [Bullet] -> [Bullet]
removeUsedBullets bullets = filter (\x -> not (bulletUsed x)) bullets

addPoints :: Int -> Player -> Player
addPoints points player = player {playerPoints = (playerPoints player) + points} 

playerKillEnemies :: GameState -> GameState 
playerKillEnemies game 
    | any (bulletsKillEnemy (playerBullets game)) (enemies game) = game {
        enemies = filter enemyAlive newEnemies, 
        playerBullets = removeUsedBullets usedBullets, 
        player = addPoints points (player game) } 
    | otherwise = game 
    where 
        usedBullets = usedPlayerBullets (enemies game) (playerBullets game)
        newEnemies = bulletsKillEnemies (playerBullets game) (enemies game)
        killedEnemies = filter (\x -> not (enemyAlive x)) newEnemies 
        points = length killedEnemies

playerOutOfScope :: Player -> Player 
playerOutOfScope player 
    | fst (playerPos player) < -475 = player {playerPos = (-475, -250), playerVel = (0, 0)} 
    | fst (playerPos player) > 475 = player {playerPos = (475, -250), playerVel = (0, 0)} 
    | otherwise = player 

updatePlayerPosition :: Player -> Player
updatePlayerPosition player 
    | (fst (playerPos player) <= -475  && fst (playerVel player) == -4 ) || (fst (playerPos player) >= 475  && fst (playerVel player) == 4) = playerOutOfScope player 
    | otherwise =  player { playerPos = (posX + velX, posY)}
    where 
        (velX, _) = playerVel player 
        (posX, posY) = playerPos player

movePlayer :: GameState -> GameState
movePlayer game = game {player = updatePlayerPosition (player game)} 

updateBulletPosition :: Bullet -> Bullet
updateBulletPosition bullet = bullet { bulletPos = (posX, posY + velY)} 
    where 
        (_, velY) = bulletVel bullet 
        (posX, posY) = bulletPos bullet 

bulletOutOfScope :: Bullet -> Bool
bulletOutOfScope bullet = snd (bulletPos bullet) > 300 || snd (bulletPos bullet) < -300 

removeBulletsOutOfScope :: [Bullet] -> [Bullet]
removeBulletsOutOfScope bullets = filter (\bullet -> not (bulletOutOfScope bullet)) bullets

updateBulletsPosition :: [Bullet] -> [Bullet]
updateBulletsPosition bullets = removeBulletsOutOfScope updatedBullets 
    where 
        updatedBullets = map updateBulletPosition bullets

movePlayerBullets :: GameState -> GameState
movePlayerBullets game = game {playerBullets = updateBulletsPosition (playerBullets game)} 

moveEnemiesBullets :: GameState -> GameState 
moveEnemiesBullets game = game {enemiesBullets = updateBulletsPosition (enemiesBullets game)} 

selectRandomEnemies :: StdGen -> [Enemy] -> Int -> ([Enemy], StdGen)
selectRandomEnemies gen _ 0 = ([], gen)
selectRandomEnemies gen enemies count = ((enemy : rest), genNew2)
  where 
    (randomIndex, genNew) = randomR (0, length enemies - 1) gen :: (Int, StdGen)
    enemy = enemies !! randomIndex
    remainingEnemies = delete enemy enemies
    (rest, genNew2) = selectRandomEnemies genNew remainingEnemies (count - 1)

addEnemiesBullets :: Float -> [Enemy] -> [Bullet]
addEnemiesBullets count enemies = fmap (\x -> Bullet {bulletPos = (enemyPos x), bulletVel = (0,((-1.5) - count*(0.5))), bulletUsed = False}) enemies 

enemiesShoot :: GameState -> GameState
enemiesShoot game 
    | shootTimer game == 0 = game { enemiesBullets = newBullets, shootTimer = shootTimer game + 1, generator = gen}
    | shootTimer game >= 1 && shootTimer game <= 40 = game {shootTimer = shootTimer game + 1}
    | otherwise = game {shootTimer = shootTimer game - 41}
    where
        numberOfEnemies = min 4 (length (enemies game))
        (selectedEnemies, gen) = selectRandomEnemies (generator game) (enemies game) numberOfEnemies
        newBullets = enemiesBullets game ++ addEnemiesBullets (countOfRound game) selectedEnemies

moveEnemyRight :: Float -> Enemy -> Enemy
moveEnemyRight deltaX enemy | deltaX > 0 = enemy {enemyPos = (posX + deltaX, posY)} | otherwise = enemy 
    where
        (posX, posY) = enemyPos enemy

moveEnemyLeft :: Float -> Enemy -> Enemy
moveEnemyLeft deltaX enemy | deltaX > 0 = enemy {enemyPos = (posX - deltaX, posY)} | otherwise = enemy 
    where
        (posX, posY) = enemyPos enemy

moveEnemyDown :: Enemy -> Enemy
moveEnemyDown enemy = enemy {enemyPos = (posX, posY - 5)} 
    where
        (posX, posY) = enemyPos enemy

updateEnemyPosition :: Float -> EnemyDirection -> Enemy -> Enemy
updateEnemyPosition deltaX currentDir enemy 
    | currentDir == RightDir = moveEnemyRight deltaX enemy
    | currentDir == LeftDir = moveEnemyLeft deltaX enemy
    | currentDir == DownDirAfterRight = moveEnemyDown enemy
    | currentDir == DownDirAfterLeft = moveEnemyDown enemy

updateEnemiesPosition :: Float -> EnemyDirection -> [Enemy] -> [Enemy]
updateEnemiesPosition deltaX currentDir enemies
    | deltaX <= 10 = map (updateEnemyPosition deltaX currentDir) enemies
    | otherwise = map (updateEnemyPosition 10 currentDir) enemies

getNextDirection :: EnemyDirection -> EnemyDirection
getNextDirection RightDir = DownDirAfterRight
getNextDirection LeftDir = DownDirAfterLeft
getNextDirection DownDirAfterRight = LeftDir
getNextDirection DownDirAfterLeft = RightDir

shouldChangeDirection :: EnemyDirection -> [Enemy] -> Bool
shouldChangeDirection currentDir enemies
    | currentDir == RightDir = any (\enemy -> fst (enemyPos enemy) >= 480) enemies 
    | currentDir == LeftDir = any (\enemy -> fst (enemyPos enemy) <= -480) enemies 
    | otherwise = True

nextDirection :: EnemyDirection -> [Enemy] -> EnemyDirection
nextDirection currentDir enemies 
    | shouldChangeDirection currentDir enemies = getNextDirection currentDir
    | otherwise = currentDir

moveEnemies :: GameState -> GameState
moveEnemies game = game { 
    enemies = updateEnemiesPosition (countOfRound game) (enemyDirection game) (enemies game),
    enemyDirection = nextDirection (enemyDirection game) (enemies game) } 

generateNewEnemies :: GameState -> GameState 
generateNewEnemies game 
    | (enemies game) /= [] = game 
    | newEnemiesTimer game < 200 = game { newEnemiesTimer = (newEnemiesTimer game) + 1} 
    | otherwise = game {enemies = seriesOfEnemies, newEnemiesTimer = 0, countOfRound = (countOfRound game) + 1}

update :: Float -> GameState -> IO GameState
update _ game
  | exit game = exitSuccess 
  | menu game  == StartBoardMenu = loadTopTen game 
  | menu game  == EndMenu && (not (playerScoreSaved (player game))) = saveAndLoadScores game 
  | menu game  == EndMenu  && (playerScoreSaved (player game)) = return game
  | menu game  == GameMenu = return $ (checkEndGame . enemiesKillPlayer . playerKillEnemies .  movePlayer . movePlayerBullets . moveEnemiesBullets . enemiesShoot . moveEnemies . generateNewEnemies) game 
  | otherwise = return game 